doc/connectivity/networking/api/net_l2.rst

```diff
@@ -123,24 +123,58 @@ here as well.  There are two specific differences however:
   IEEE 802.15.4 device driver must provide a valid pointer on such
   relevantly filled-in API structure.
 
-- Sending a packet is slightly different than in Ethernet. IEEE 802.15.4 sends
-  relatively small frames, 127 bytes all inclusive: frame header,
-  payload and frame checksum.  Buffers are meant to fit such
-  frame size limitation.  But a buffer containing an IPv6/UDP packet
-  might have more than one fragment. IEEE 802.15.4 drivers
-  handle only one buffer at a time.  This is why the
+- Sending a packet is slightly different than in Ethernet. Most IEEE 802.15.4
+  PHYs support relatively small frames only, 127 bytes all inclusive: frame
+  header, payload and frame checksum. Buffers to be sent over the radio will
+  often not fit this frame size limitation, e.g. a buffer containing an IPv6
+  packet will often have to be split into several fragments and IP6 packet headers
+  and fragments need to be compressed using a protocol like 6LoWPAN before being
+  passed on to the radio driver. Additionally the IEEE 802.15.4 standard defines
+  medium access (e.g. CSMA/CA), frame retransmission, encryption and other pre-
+  processing procedures (e.g. addition of information elements) that individual
+  radio drivers should not have to care about. This is why the
   :c:struct:`ieee802154_radio_api` requires a tx function pointer which differs
-  from the :c:struct:`net_if_api` send function pointer.
-  Instead, the IEEE 802.15.4 L2, provides a generic
-  :c:func:`ieee802154_radio_send` meant to be given as
-  :c:type:`net_if` send function. It turn, the implementation
-  of :c:func:`ieee802154_radio_send` will ensure the same behavior:
-  sending one buffer at a time through :c:type:`ieee802154_radio_api` tx
-  function, and unreferencing the network packet
-  only when all the transmission were successful.
-
-Each IEEE 802.15.4 device driver, in the end, will need to call
-``NET_DEVICE_INIT_INSTANCE()`` that way:
+  from the :c:struct:`net_if_api` send function pointer. Zephyr's native
+  IEEE 802.15.4 L2 implementation provides a generic :c:func:`ieee802154_send`
+  instead, meant to be given as :c:type:`net_if` send function. The implementation
+  of :c:func:`ieee802154_send` takes care of IEEE 802.15.4 standard packet
+  preparation procedures, splitting the packet into possibly compressed,
+  encrypted and otherwise pre-processed fragment buffers, sending one buffer
+  at a time through :c:type:`ieee802154_radio_api` tx function and unreferencing
+  the network packet only when the transmission as a whole was either successful
+  or failed.
+
+Interaction between IEEE 802.15.4 radio device drivers and L2 is bidirectional:
```

Comment:
@rlubos This is basically the same simple and already existing concept I would propose in the RFC we discussed to achieve better separation of L1/L2 concerns for the CSL feature and other things that have been introduced into the nRF5 driver for Thread 1.2+ support.

This would make it possible to comply with the standard which clearly (and IMHO rightly so) places IE-Handling for CSL (and also some other things that are now inside the nRF5 driver) into L2. At the same time we'd be able to keep the necessary hooks inside the driver for close interaction between L1 and L2 for performance optimization and timing sensitive tasks w/o compromising encapsulation, standard compliance and reusability.

This would make it so much easier for other driver maintainers to close the Thread 1.2+ gap in their drivers which would be great for additional Thread 1.2 protocol acceptance, I think. RAW and native L2 would get no-ops for these functions for the time being so that no extra work is required for these stacks. They simply wouldn't support those features.

What do you think? Does it make sense to pursue this idea further in an RFC together with Nordic's OT team?

Reply:
@tbursztyka What's your opinion as IEEE 802.15.4 maintainer?

Reply:
> This is basically the same simple and already existing concept I would propose in the RFC we discussed to achieve better separation of L1/L2 concerns for the CSL, general IE handling, key management and secure ACK features plus other minor things that have been introduced into the nRF5 driver for Thread 1.2+ support.

@edmont @canisLupus1313 @czeslawmakarski @jciupis @hubertmis It would be good to get some early feedback whether such model would be a fit for OpenThread radio platform wrt advanced features mentioned above.

Reply:
@fgrandel Although I can't provide any definite feedback until there's a specific RFC that can be discussed, I can't think of a way to implement the advanced features you've mentioned in L2 without compromising performance of Nordic solutions significantly. As an example, consider the dual-core nRF5340 SoC where the 802.15.4 radio driver is running on a different core than the Zephyr ieee802154 API is exposed on. The way I understand your idea, it would effectively require us to move the implementation of certain features from the nRF5340 network core to the application core. The latency introduced by serializing calls between the cores is enough to make Ack handling in L2 not feasible, i.e. the overall time penalty is larger than AIFS defined in the IEEE 802.15.4 spec. Other features could maybe be moved and still be functional, but the performance drop caused by such changes would not be acceptable. That's why they were implemented in the driver in the first place.

Perhaps it would be possible to introduce new capabilities in the API that indicate if the driver supports these advanced features and provide a software, L2 alternative that can be used instead. What do you think?

Reply:
Thanks @rlubos to start off the discussion re appropriate architectural L1/L2 interface patterns. I'll preliminarily respond to @jciupis here to see whether we get the discussion going in a constructive way and set the adequate tone for a vendor-neutral open-source Linux Foundation platform. But then it's probably a good idea to really start an RFC now, as @jciupis also suggests. I'll take care of that.

@jciupis Thanks for taking the time to look into this. No, you misunderstand me that I want to take hardware ACK away from the nRF5 driver and yes, **you're totally right that it must be possible to offload performance critical L2 features to specialized peripherals or MCUs** as well as provide hooks to selectively offload L2 features to vendor/hardware-optimized proprietary code if driver maintainers choose to do so. That's why I always carefully distinguish between what I call "hard MAC" (or offloading) and "soft MAC" (or as Zephyr likes to call it: "L2 stack") implementation approaches.

Please be aware that this whole discussion is strictly not required for what I personally need. I could easily extend the radio API in a similarly protocol-specific way for TSCH as the OT team did for Thread. But I deeply believe that such an approach would be highly detrimental to the overall community appeal, future uptake and long-term success of Zephyr's IEEE 802.15.4 stack. That's the only reason why I'm challenging the current approach.

So yes, I'll be available to jointly contribute to an RFC with you. But I'll only continue to invest my personal time if we can agree on some shared goals and principles of collaboration. I don't like to publicly fight windmills or put my own little weight against the economic interests of a large vendor (which I otherwise admire for tons of value it already provides to me and the Zephyr community).

Here is what I find necessary for us to collaborate in a constructive way:

1. First we need to agree that our main goal will be to achieve **vendor- and protocol-neutrality** of all of the platform's shared APIs (Kconfig, devicetree, net_mgmt, radio API, etc) in the common interest of the whole being much more than its parts.
2. We also need to agree that such an approach requires us to combine good design with pragmatism, so that while we open many doors, **we only solve problems that we really have to deliver on our respective requirements**.
3. Then we need to agree that the easiest way to remain vendor and protocol agnostic in a community-maintainable way is to **strictly stick to the standard's well-defined service primitives and data structures**. This is the most important point. I'm definitely NOT available to discuss any OT or nRFx quirks in the same way as you probably don't want to involve yourself in Contiki's way of implementing TSCH or TI SubG driver internals. The standard is our common ground, so everything we propose should refer to a specific section of the IEEE 802.15.4-2020 standard. Otherwise this discussion will never scale to a point that I'm able to support it.

In other words: Can we agree that the fact that Nordic and the Open Thread team have decided to offload a small subset of the standard's L2 features into a specialized MCU is legitimate but shouldn't force everyone else having to do the same plus it should be done in a way that enables re-use of shared features in other protocols?

Of course your own point is equally important and implicitly already included in 1) - offloading of L2 features to a vendor's hardware must of course be possible. **Feel free to add your own points.** I think we're together on Zephyr because we fundamentally share the same collaborative approach.

Now if we can agree on these IMO self-evident points, let's talk solutions, please - because **vendor-neutrality, re-usability and protocol-agnosticism is unfortunately no longer the current state of affairs in the IEEE 802.15.4 radio API as I'll show in more detail further down**.

The rest of this post is just to start the discussion off and will (in the case you decide that collaboration based on these principles makes sense) mostly be copied into the RFC. **If you don't agree to the above principles: no need to read on. In this case please let me know** so that I stop wasting my time.

# Why collaborate?

1. The reason why I bring this whole discussion up is that **I'm trying to implement a standard protocol (TSCH) with a standard PHY (SubG SUN-FSK, channel page 9) on Zephyr-supported hardware (TI's CC13/26xx)**. While doing so I looked into the current radio API and found that comparatively recent architectural choices stand fundamentally in the way of such an IMHO totally legitimate endeavor. In the best case (e.g. CSL, key management, secure ACK) the current radio API requires massive code duplication across drivers (e.g. CC13/26xx) as well as in Zephyr's native L2 stack to be supported w/o any relevant contribution to the specific needs of TSCH although most of the required underlying standard primitives and data structures are exactly the same and could be provided in a 100% re-usable way. In the worst case the current radio API isn't even basically compatible with the IEEE 802.15.4 standard and therefore not extensible or re-usable in a sensible way at all (e.g. frame counter management or - to be fair and name an example from another hardware platform - SubG channel page selection in the TI driver).

2. But then there is another problem which is totally unrelated to TSCH and IMHO needs urgent fixing: **The effort currently required to close the Thread 1.2+ implementation gap in other vendors' drivers is enormous and much higher than it needs to be. Plus the radio API urgently needs to support alternative hardware offloading choices if we want any other Zephyr driver maintainer to support recent versions of (Open)Thread in Zephyr.** This is IMO just not realistically possible right now. We do not want to provide good arguments to vendors who sit on the fence to turn their back to supporting IEEE 802.15.4 and Thread in Zepyhr, right? How did it happen in the first place that only one of the project partners officially supports Thread in Zephyr? I honestly don't know, do you? If these other vendors join forces around another Thread or IEEE 802.15.4 implementation elsewhere then Zephyr has a big problem, I suppose. On the other hand: If we make Thread 1.2+ support so easy in Zephyr that community maintainers can do it - even without official vendor support - then the remaining features of Zephyr will be extremely hard to match by any alternative framework no matter how much proprietary effort goes into it. This analysis may be totally naive, though. Let me know if you think this is all just wishful thinking.

# How can it be done?

There are established architectural patterns in the radio API that IMO can be made fruitful to guarantee that the radio API returns to be fully vendor- and protocol-agnostic with a lot of re-use across Thread and TSCH and across drivers while of course enabling (but not enforcing) the type of vendor-specific offloading that you have in mind. 

Here an obviously non-exhaustive list of **existing architectural patterns that play a role**:

1) Enable **optional offloading of L2 features to hardware or proprietary firmware** wherever necessary: It was always the case that the radio API allowed for offloading but ensured that hardware-specific features were either strictly optional (e.g. promiscuous mode, alternative PHYs and frequency bands, the number of supported channels, etc.) or - for mandatory features from a standard perspective - there was always a choice between "soft MAC" and "hard MAC" implementation (e.g. automatic ACK in driver firmware or hardware, CSMA/CA, timed TX/RX, etc.).

2) **Callbacks to maintain clean L1/L2 separation of concerns w/o compromising performance/timing**: Where close "reverse" interaction between L1 and L2 is required due to strict timing specs, performance requirements or simply to inform upper layers of HW events (which is unavoidable even when offloading features), a promising L1->L2 callback pattern was already pioneered in Zephyr's radio API and - more importantly - has been scaled and battle tested for a long time in Linux's IEEE 802.15.4 and bluetooth stacks. This is what I documented and highlighted in this PR. Such a simple callback / "inversion of control" pattern allows for re-usable callouts from the driver into the MAC layer w/o compromising on standard compliance, re-usability and encapsulation. There is definitely no need to have an IE parser and writer in all drivers AND in L2 just to support time critical time stamping of IEs or make information from e.g. PHY-specific IEs available to drivers while having MAC-specific protocols access "their" IEs independently. Such a resource hungry approach would quickly become an unmaintainable mess with loads of inconsistencies and bugs across drivers and L2s.

3) **Design based on standardized granular service primitives and data structures**: I've gone to great lengths in some of my recent PRs to ensure and demonstrate that the native stack is fully compliant with granular IEEE 802.15.4 service primitives and data structures. This was not hard as this was mostly already the case just not documented everywhere and for the rest I'm closing the gap quickly now in my upcoming PRs.

This last point will be very important for potential discussions in an RFC and I must be sure that you're willing to stick to the standard's primitives, data structures and naming principles all of the time. **IMO our whole discussion must be strictly limited to gain a common understanding of which subset of service primitives and data structures needs to be supported right now in Zephyr and then how to map this subset to appropriate shared APIs plus identifying and supporting typical hardware offloading choices for them.** As the Thread standard has been very carefully based on standard IEEE 802.15.4 service primitives and data structures plus TSCH _is_ the standard, I don't see any problems with such an approach in practice at all. Most existing stacks do just that.

**What we should and can definitely fix with such an approach is the current monolithic and/or restricted way in which features have been implemented in the API**: CSL support, information elements, key management, enhanced ACKs, frame counter management, etc. can all easily be converted to be based on re-usable service primitives and data structures from the IEEE 802.15.4 standard. Then we need a thin OT-to-primitives adaptation layer plus the nRF driver needs to selectively offload whatever bundle of these primitives internally while keeping the door open for other vendors who prefer a "soft MAC" solution without anyone having to implement it of course unless actually needed.

Can you see how such an approach will enable you to continue to build your driver on whatever vendor-specific solution you like? If other vendors or community driver maintainers do not have their own ready-made IE parser at hand, then we have a highly re-usable design ready that lowers the hurdle for massive collaboration which is the strength of Open Source after all! I think that's how Zephyr as a joint community should look like! :-)

Some of this needs to be fixed by the OT and nRF driver teams (e.g. writing adaptation layers for OT and the nRF driver), some of it can be contributed by myself (e.g. certain generic IE parsing/writing primitives and time stamping support as well as further improved support for generic RX windows and timed TX plus the TI CC13/26xx equivalent mapping of standard primitives to TI hardware quirks).

If you agree that it makes sense to collaborate based on these fundamental assumptions, I'll take the next step and set up an RFC for us to collaborate - copying in most of this answer as a starting point.

Reply:
@fgrandel thank you for taking the time to prepare a detailed, in-depth reply.

I misunderstood your original proposal and assumed you're suggesting to enforce handling L2 features in "soft MAC". It's good to hear that we agree that this is not feasible. It is necessary to keep the door open for vendor-specific drivers to selectively offload performance critical L2 features.  

I need to discuss internally several issues and suggestions you mentioned. Let me come back with a more substantial reply in a couple of days.

Reply:
Resolved this conversation as it is not in the way of merging this PR. The conversation will be moved to an RFC as soon as jciupis returns with a reply.

Reply:
@fgrandel I've discussed the ideas you presented above internally. I think we are on the same page regarding the architectural benefits such rework of the Zephyr's IEEE 802.15.4 stack would bring. The points you raised are valid and from technical standpoint it would be a valuable improvement. However, management has not approved that we invest time in this improvement now. Our collaboration will therefore be limited to the scope of review and PR-related discussions as it has been so far, not an in-depth design work and restructuring the existing implementation.

I hope it's not a limiting factor for your work. Thank you for starting this discussion. Looking forward to more reviews and discussions in your PRs.

Reply:
@jciupis Thanks for coming back to me. No, it's not a limiting factor for me. As you approve the architectural benefits as such I'll open up an RFC as soon as I introduce changes in that respect so the context is well documented. I'll try to make my own additions in a way that will hopefully improve the radio API's re-usability and lower the bar for driver maintainers as much as my resources allow me to. Having your support in form of reviews and discussions is very welcome.

Just one extra thing: If you intend to introduce more changes to the radio API, please let's align on those changes beforehand if anyhow possible (just the API part of it) so that we don't step on each other's toes. Hope that's acceptable for you?

Reply:
> Hope that's acceptable for you?

Yes, sure.

Reply:
Just to give you an expression in which direction I'm trying to develop the API in practice: #59245 is a first effort to gently move the radio API back to standard definitions.

Similar efforts will follow for at least:
* access to the radio timer,
* timed TX
* RX windows.

I'll also propose to deprecate everything in the radio API relating to the monolithic CSL solution and provide an alternative IE handling mechanism with hardware offloading hooks that is based on standard primitives and re-usable across drivers.

These are all areas where TSCH is based on the same basic primitives as CSL.

Reply:
> Similar efforts will follow for at least:
> 
> * access to the radio timer,
> * timed TX
> * RX windows.

I like this plan very much to have vendor-agnostic APIs in Zephyr for these features.

I would extend the list with Enh-Ack content and with HW MAC encryption, which is especially useful with Enh-Acks. I believe that will be needed for TSCH.

All of this API alignment might require changes in OpenThread integration or in OpenThread itself. But I think it would be good if OpenThread supported a similar, generic, vendor-agnostic API for its integration with radios.

Reply:
@hubertmis Thanks for your encouragement - very much appreciated! :-)

The following answer is not just for you but helps me to further evolve the subject as I need to find solutions for the topics you mention anyway.

**Enh-Acks** are on my list as they are heavily used in TSCH for time synchronization (including HW assisted field injections just like in CSL and/or secure ACK IEs).

**Encryption with multiple keys** (HW assisted or not) is optional for TSCH and I personally don't need it. It just happens to be mandatory for a certain OT version. I'll not be able to contribute to that, although I agree that it should be (re-)modeled in the radio API in a way that is open for both, software and hardware assisted solutions, with frame counter injection independent of OT requirements, specific frame types or IEEE 802.15.4 sub-protocols. If someone else did this, TSCH would immediately benefit - that's the kind of potential for community synergies we should try to establish.

We also need to be careful that **enhanced ACKs are more than secure ACKs** - it's a generic frame format used in all recent IEEE 802.15.4 sub-protocols that needs to support any kind of IE and frame payload. Plus IEs, frame payload and key management are not restricted to ACKs but used in version 3 data frames, beacons, commands, etc. and must therefore be modeled as re-usable standard primitives completely independently of ACKs or CSL.

Concepts like "content" or "data" (as the current radio API unfortunately calls OT specific header IEs in ACKs) are either not well-defined for enhanced frames or ambivalent ("content" is a specific field in command frames).

For everything related to **enhanced frames (including but not limited to Acks)**, it is important from a standard's perspective to **let the API distinguish independently from any protocol or frame type between**:
- **header IEs** (not encrypted)
- **payload IEs** (encrypted)
- **frame payload** (which Enh-Ack frames also need to support)
- plus special rules for beacon and command fields of earlier frame versions
see IEEE 802.15.4-2020, sections 9.2.2 and 9.3.5.3
- **pointers to well-defined fields in IEs for HW assisted field injection**

All these concepts have already been used implicitly in the current radio API for CSL, enhanced ACKs and HW assisted encryption but in a way that cannot be re-used although clearly defined as highly re-usable primitives in the core standard.

I believe that if we manage to find a good re-usable solution for HW assisted injection of timestamps into time synchronization IEs for TSCH, this will be so similar to the HW assisted solution required for CSL and enhanced Acks that it will be available as an **easy-to-implement initial template** for those areas thereby **lowering the hurdles to further improvement** even if not implemented immediately. At least this should be our stated goal.

Fortunately the standard provides very good primitives for such an API: **We basically only have to stick to the pre-defined signatures of MLME service calls (see section 8.2)** with common fields drawn together into common struct headers for re-use across calls. So this shouldn't be too hard.

I already have [some code for Enh-Ack](https://github.com/fgrandel/zephyr/commit/3e202615f61c9da3b91f78caa14b859331f42ee8#diff-00ac646417b83c5a386b286f875c502f9b304bc8d2fe5dad27e016a48e6d0c04R1785) and [Enh-Beacons](https://github.com/fgrandel/zephyr/commit/3e202615f61c9da3b91f78caa14b859331f42ee8#diff-00ac646417b83c5a386b286f875c502f9b304bc8d2fe5dad27e016a48e6d0c04R1900) showing some initial soft-MAC IE handling (but yet w/o the HW-assisted injection support) on my public dev branch if anyone likes to have a preview. You may have to open the diff for `ieee802154_frame.c` and look out for `ieee802154_create_enh_ack_frame()` and `ieee802154_create_enh_beacon()` if the link doesn't work).

On a side note: I prefer not to call the current API "vendor specific" in a strict sense in that it leaks vendor-specific hardware concepts or something. It's just so much monolithically centered around OT's eclectic requirements bundle and a single vendor's "hard MAC" vs. "soft MAC" architectural choices, that it builds up high economic hurdles ("technical debt") in practice for alternative driver and protocol contributions. This contrasts with the typical synergies of an OSS community effort providing standard-based primitives that makes it easier rather than harder for everyone to build upon them.

Reply:
> Concepts like "content" or "data" (...) are either not well-defined for enhanced frames or ambivalent

I was imprecise. I meant IEs (both header and payload) and Frame Payload. I agree we should use unambiguous terminology.

> Fortunately the standard provides very good primitives for such an API

The standard defines very good primitives for API above L2. However, I'm not aware of good primitives at the boundary between drivers and L2.

Usually 802.15.4 stacks are implemented similarly to Wi-Fi with vendor-agnostic "upper MAC" (like Zephyr's L2 or OpenThread) and vendor-specific "lower MAC" (or drivers) implementing timing-critical MAC features like Acks, often running on a separated CPU or in hardware.

I agree we should try to be aligned with 802.15.4 primitives defining 802.15.4 driver API in Zephyr, but I'm afraid the standard won't solve all problems, because it does not expect splitting MAC to lower and upper parts.

I'm happy to review and discuss new driver APIs and the modification of existing ones.

Reply:
> The standard defines very good primitives for API above L2. However, I'm not aware of good primitives at the boundary between drivers and L2.

@hubertmis  Nice, very good points that immediately made me think more. 100% agreed that the MAC <-> PHY API has very little defined structure in the standard. It just has a name and defines a few PHY PIB attributes. And of course you're right that this is what a "radio API" is about. That's where I was really imprecise.

I think what confuses us is that **I was mixing up two things**:
1) aspects of a re-usable and vendor-agnostic **MAC <-> PHY API** which legitimately may be called "radio" API, which has no detailed description in the standard and which you certainly have it in mind
2) aspects of a **hard MAC API** that allows for selective offloading of MAC features to hardware and where speaking of implementing (parts of) MLME would be valid

The fact that these aspects are equally mixed up in the current "radio" API (which it no longer is) tells me that we should keep those concepts very well apart from now on to avoid further confusion.

**I therefore propose the following naming convention**:
* **radio API**: this name should be **strictly reserved to the software wrapping a low-level radio peripheral or transceiver**, it should be vendor-agnostic (not because the standard defines it but because Zephyr is a vendor agnostic platform, MLME plays no role here, PHY PIB attributes do)
* **hard MAC API**: this name should be **strictly limited to (optional) offloading of MAC features to hardware** (and based on MLME concepts wherever possible to remain equally vendor agnostic)
* **driver API: the combination of radio and hard MAC API.** This is what a vendor needs to implement to support a given piece of hardware and which should be open to implementation variants (soft MAC vs. offloading) depending on the vendor's specific hard/soft MAC architectural choices.

I don't like the word "lower MAC" as a synonym for "driver" as you propose although I agree this is common. It is confusing for low-level radio stuff which has nothing to do with MAC. And it is confusing in the context of hardware that implements a "modem" approach with a full featured "upper MAC in hardware" which still needs to be wrapped in a vendor-agnostic driver API.

**The radio API should be very simple and effortless to implement** so that "entry level" support for all MAC features via soft MAC is really, really cheap. A single community contributor even if not backed by a vendor should be able to implement this API in no time for a new SoC or transceiver so that many more chips can support Thread and/or TSCH in the future. Things like timed TX, RX windows and radio timer access belong here (but neither Imm-Ack nor Enh-Ack, no CSL, no IEs, etc.).

**The hard MAC API should enable vendors with more resources and/or elaborate hardware architectures to (optionally) offload as much as they like while still being vendor agnostic.** This allows vendors to differentiate their contribution with superior performance and best-of-class support for vendor-specific hardware features. Things like ACK (enhanced and immediate), IE handling, CSL, management of keys, frame counter handling, etc. belong here. None of this should be strictly required for basic Thread/Matter and/or TSCH support.

**I think this should cover the diverse needs of a large spectrum of Zephyr community stakeholders, do you agree?**

Upper layers will speak to a single API that is fully MLME compliant (a combination of Zephyr's L2 agnostic L2 <-> L3 API plus L2-specific net_mgmt(), devicetree and Kconfig stuff). Then each L2 implementation (currently OT vs. native) decides depending on the driver's generically advertised implementation choices whether to delegate a MAC feature to the driver's hard MAC API or whether to provide it itself calling into the driver's low-level radio API instead.

I think these concepts should be pretty much in line with Linux architecture in similar cases which is where we can learn quite a bit when it comes to being vendor agnostic, right?

Reply:
Hi @hubertmis , I agree that dividing up concepts based on "optional" vs. "mandatory" is useful in certain contexts. But I think its not the right way to divide up the IEEE 802.15.4 API for two reasons:

1) Optional vs. mandatory is relative to application requirements and depends on your build and runtime configuration. So structurally dividing up the API along these lines is not practical IMO as **the distinction is not fully known at design time**.

**Example 1**: In the context of PHYs with high symbol rates it may be mandatory for a driver to support ACK in hardware for full spec compliance. In PHYs with low symbol rates, however, ACK can easily be implemented in software and is then optional in the driver.

**Example 2**: A timing sensitive protocol as TSCH which will not work w/o a low-level radio timer in hardware (timed TX/RX) for intra-slot timing which is then a mandatory radio API feature for that application profile albeit a rather minimal one.

**Example 3**: If you have a low energy, broadcast-only ALOHA/no ACK sensor node with a proprietary non-IP protocol above it then you need no hard MAC API support in your driver and even some of the most basic radio API features like CCA and ED are then optional.

2) **Radio API and hard MAC API belong to different architectural layers (L1 vs. L2) that must not be mixed up at all. The optional/mandatory list freely mixes L1 and L2 concerns, however, as does the current IEEE 802.15.4 so-called "radio" API.**

Disentangling this mixture of L1 and L2 concepts in the API is IMO the basic problem that we need to solve in the first place. It's very important that we can agree on this as laid out in my very first post.

> drivers implement only a part of MAC, so 802.15.4 MLME primitives usually do not fit well

For the moment being I'd be satisfied if we just cleanly distinguished between L2 (MAC) and L1 (radio) features by simply looking them up in the standard. Sections 6-9 in IEEE 802.15.4-2020 are L2/MAC features, everything in chapters 10+ are L1/radio/PHY features. That should be easy enough.

If you like go through your list and label it with sections from IEEE 802.15.4-2020 and I'm sure you'll arrive at a much better L1/L2 divide plus at a much finer granularity (i.e. level of re-usability). That's what we agreed upon as our approach and as a precondition for any discussion re IEEE 802.15.4 API anyway, right?

Maybe we're just talking about different things? Let me take just one example from your list to make it clear what I mean by "built on standard primitives" (the same analysis can and must be done for the rest of your list if we want to make it re-usable):

> Auto Imm-Ack transmission (requires the capability of setting the frame pending bit)

First of all: Everything related to Imm-Ack is L2/MAC level (defined in sections 6.7.4, 7.3.3. etc.) so it must be part of the hard MAC API of the driver if optionally implemented in hardware (with a soft MAC fallback outside drivers of course for the rest of the world).

But then **things related to the "frame pending" bit are unrelated and orthogonal to the Imm-Ack/Enh-Ack distinction** and even to any specific frame type. Section 7.2.2.4 gives an overview over all use cases in which the frame pending bit may play a role and it is not a co-incidence that the frame pending bit is specified in the general frame control field (section 7.2.2.4) rather than for ACKs only. So tying frame pending in any way to ACKs (or even worse to CSL) is just not re-usable enough as it breaks the structure defined in MLME primitives.

Even in the context of CSL, only the coordinator (sender) needs to implement a pending transactions list (section 6.12.2.3). The CSL destination (sleepy end device, confusingly also called "CSL coordinator" in the CSL context) will never need indirect transmission as its counterpart is "always on". So why make it mandatory?

**Interaction with the MAC's data TX/RX service (aka MCPS-DATA.*, see section 8.3) must be totally unaware of the underlying protocols** be it CSL, TSCH, RIT, LE transmissions, plain beacon enabled or non-enabled PANs, encryption, etc. Trying to provide several individual TX/RX related API calls for each combination of protocols in the way CSL or secure ACK was implemented would mean a combinatorial explosion that will never scale. It is therefore completely non-standard and against MLME definitions that the current offloading API requires CSL-/ACK-specific interactions on a per-frame level.

The correct way to do this instead is that **the soft or hard MAC exposes the necessary MAC PIB attributes via configure calls that are in fact MCPS-SET/GET primitives** (see section 8.2.6) and which set the protocol(s) to be used for the PAN for all subsequent TX/RX calls (see section 8.4.3.2) which are then to be handled transparently by the (partially offloaded) MAC layer.

This simplifies usage of the driver API because many details simply disappear from it:

* Whether the frame pending bit is to be set or not must be of no concern to the caller. When, where and how the frame pending bit is to be set depends on the protocols enabled in the PAN alone (see section 7.2.2.4). It must be set automatically on certain TX frames by the offloaded hard MAC. This happens when the standard call's "indirect transmission" flag was set and the transaction therefore landed in the pending transaction list which must be implemented in hard MAC in the offloading case (MCPS-DATA.request, section 8.3.2).

* Incoming ACKs (Imm AND Enh) will NOT be signaled separately by the driver API but through a single MCPS-DATA.confirm primitive including among many other things the incoming frame pending bit. Nothing in the world keeps us from defining a hard MAC driver API that does just that. This can then be used in CSL but it can equally be used in any other context and of course a soft MAC counterpart can easily be provided as well. MCPS-DATA.confirm is very close to our ieee802154_handle_ack() - frame pending is included here.

**All remaining OT-specific interaction between offloaded and non-offloaded parts of the MAC is to be handled via OT- and protocol-specific hard MAC-to-soft MAC callbacks to be implemented by OT-specific adapters and that the driver API (and community driver maintainer) is totally unaware of.** Examples are OT-specific frame counters handling or low-level timing info that is then interpreted as CSL phase or rendezvous timepoint in the non-offloaded part of OT. This is where all kinds of OT idiosyncrasies can be hidden from the rest of the world if required.

What I'm saying here amounts to the fact that **hard MAC must be to some extent aware of the chosen L2 stack (native vs. OT)** in certain strictly limited ways that can probably even be hidden behind L2-agnostic common defines or external functions. But this is much, much less than forcing knowledge of the chosen L2 stack or offloading choices to all callers of the driver API! **If a community driver maintainer chooses not to offload anything such details must be completely irrelevant.**

This is possible because the same OT-specific callbacks can be implemented in an OT-specific soft MAC adapter which the OT stack will have to provide as a fallback for community drivers that do not offload CSL. **This fallback layer must be tuned to work with basic timed/timestamped TX/RX only which is very easy to implement for community driver maintainers** and is all that is really required for CSL (or TSCH) support in the soft-MAC case.

> I think the "optional capabilities" list does not include services defined as MLME primitives.

I hope that I've just proven the contrary - it just requires much more detail than just jotting down a list of features.

But if it's still not clear what I mean then I'd prefer to discuss alternative PRs from now on rather than arguing academically back and forth, ok?

Reply:
Yes, i didn't fully understand your proposal. I feel we are trying to achieve the same goal but in different ways. I agree we should discuss examples on PRs or RFCs

Reply:
...and here is the first [PR](https://github.com/zephyrproject-rtos/zephyr/pull/60343) that allows us to discuss almost all of the above issues based on a simple real-world example.